### PR TITLE
[PDI-15652] - Several steps / jobs don't work with encrypted password…

### PR DESCRIPTION
**Pull request by YuryBY**
… in a parameter

### CODE REVIEW

**Comment by YuryBY** [comment]
@pentaho-nbaker, @mbatchelor, @pamval, could you please take a look?

**Comment by YuryBY** [comment]
@pentaho-nbaker, @mbatchelor, @pamval, any updates?

File: engine/src/org/pentaho/di/job/entries/sendnagiospassivecheck/JobEntrySendNagiosPassiveCheck.java [hunk 1]
```diff
@@ -548,6 +550,18 @@ public Result execute( Result previousResult, int nr ) {
     return result;
   }
 
+  @VisibleForTesting
+  protected String resolvePassword( String password ) {
```
Comment (mbatchelor) [inline engine/src/org/pentaho/di/job/entries/sendnagiospassivecheck/JobEntrySendNagiosPassiveCheck.java]: So much duplication. Please move this method into Utils class as a static helper method.

File: engine/src/org/pentaho/di/job/entries/sftp/JobEntrySFTP.java [hunk 2]
```diff
@@ -756,6 +758,18 @@ public Result execute( Result previousResult, int nr ) {
     return result;
   }
 
+  @VisibleForTesting
+  protected String resolvePassword( String password ) {
```
Comment (mbatchelor) [inline engine/src/org/pentaho/di/job/entries/sftp/JobEntrySFTP.java]: As above

File: engine/src/org/pentaho/di/job/entries/sftpput/JobEntrySFTPPUT.java [hunk 3]
```diff
@@ -987,4 +989,16 @@ public void check( List<CheckResultInterface> remarks, JobMeta jobMeta, Variable
         AndValidator.putValidators( JobEntryValidatorUtils.integerValidator() ) );
   }
 
+  @VisibleForTesting
+  protected String resolvePassword( String password ) {
```
Comment (mbatchelor) [inline engine/src/org/pentaho/di/job/entries/sftpput/JobEntrySFTPPUT.java]: Ditto

File: engine/src/org/pentaho/di/job/entries/getpop/JobEntryGetPOP.java [hunk 4]
```diff
@@ -716,8 +716,19 @@ public String getPassword() {
...
   }
 
+  /**
+   * @return Returns resolved decrypted password or null
+   * in case of {@link #getPassword()} returns null.
+   */
   public String getRealPassword() {
-    return environmentSubstitute( getPassword() );
+    String resolvedPassword = environmentSubstitute( getPassword() );
+    if ( resolvedPassword != null ) {
```
Comment (mbatchelor) [inline engine/src/org/pentaho/di/job/entries/getpop/JobEntryGetPOP.java]: As noted below, this code should be in a Utils method ( core/org.pentaho.di.core.util.Utils I believe) and then used everywhere.

File: engine/src/org/pentaho/di/trans/steps/exceloutput/ExcelOutput.java [hunk 5]
```diff
@@ -855,4 +857,16 @@ private String reEncodeWriteAccessIfNecessary( String writeAccess ) {
     }
     return writeAccess;
   }
+
+  @VisibleForTesting
+  protected String resolvePassword( String password ) {
```
Comment (mbatchelor) [inline engine/src/org/pentaho/di/trans/steps/exceloutput/ExcelOutput.java]: Same thing

File: engine/src/org/pentaho/di/trans/steps/excelwriter/ExcelWriterStep.java [hunk 6]
```diff
@@ -971,4 +973,16 @@ protected void protectSheet( Sheet sheet, String password ) {
       sheet.protectSheet( password );
     }
   }
+
+  @VisibleForTesting
+  protected String resolvePassword( String password ) {
```
Comment (mbatchelor) [inline engine/src/org/pentaho/di/trans/steps/excelwriter/ExcelWriterStep.java]: Same thing

File: engine/src/org/pentaho/di/trans/steps/httppost/HTTPPOST.java [hunk 7]
```diff
@@ -494,7 +495,8 @@ public boolean init( StepMetaInterface smi, StepDataInterface sdi ) {
       data.realProxyHost = environmentSubstitute( meta.getProxyHost() );
       data.realProxyPort = Const.toInt( environmentSubstitute( meta.getProxyPort() ), 8080 );
       data.realHttpLogin = environmentSubstitute( meta.getHttpLogin() );
-      data.realHttpPassword = environmentSubstitute( meta.getHttpPassword() );
+      data.realHttpPassword =
+          Encr.decryptPasswordOptionallyEncrypted( environmentSubstitute( meta.getHttpPassword() ) );
```
Comment (mbatchelor) [inline engine/src/org/pentaho/di/trans/steps/httppost/HTTPPOST.java]: This is what I expected everything else to look like - see my comment on the PDI-15652. So if it's OK to do it simply and inline like this here, why can't everything else be this simple? so why the resolvePassword method added to stuff above, but this one is OK to be a single statement? I prefer the single statement.

File: engine/src/org/pentaho/di/trans/steps/monetdbbulkloader/MonetDBBulkLoader.java [hunk 8]
```diff
@@ -735,4 +737,15 @@ protected static void executeSql( String query, String host, int port, String us
 
   }
 
+  @VisibleForTesting
+  protected String resolvePassword( String password ) {
```
Comment (mbatchelor) [inline engine/src/org/pentaho/di/trans/steps/monetdbbulkloader/MonetDBBulkLoader.java]: One way or the other here.

**Comment by YuryBY** [comment]
@mbatchelor, the PR is updated. Thanks.

**Pull request merged**